earth_enterprise/src/fusion/autoingest/AssetRegistry.h

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2020 The Open GEE Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef ASSET_REGISTRY_H
+#define ASSET_REGISTRY_H
+
+#include <functional>
+#include <memory>
+#include <unordered_map>
+
+template <class AssetType>
+class AssetRegistry{
+public:
+  AssetRegistry() = delete;
+
+  class AssetPluginInterface {
+    /*
+    * Collection of plugin-specific versions of functions and data.
+    */
+    public:
+      using new_from_dom_func = std::function<std::shared_ptr<AssetType>(void*)>;
+      using new_invalid_func = std::function<std::shared_ptr<AssetType>(const std::string &ref)>;
+      new_from_dom_func pNewFromDom;
+      new_invalid_func pNewInvalid;
+
+      explicit AssetPluginInterface(new_from_dom_func pNFD, new_invalid_func pNI):
+        pNewFromDom(pNFD), pNewInvalid(pNI) {}
+  };
+
+
+  class PluginRegistrar
+  {
+    /*
+    * PluginRegistrar objects are instantiated within a plugin's cpp file to
+    * register the plugin.
+    */
+    public:
+      explicit PluginRegistrar(
+        std::string const & name, 
+        std::unique_ptr<AssetPluginInterface> assetPlugin)
+      {
+          AssetRegistry<AssetType>::RegisterPlugin(name, std::move(assetPlugin));
+      }
+  };
+
+  static void RegisterPlugin(
+    std::string const & assetTypeName, 
+    std::unique_ptr<AssetPluginInterface> assetPlugin)
+  {
+    PluginRegistry()[assetTypeName] = std::move(assetPlugin);
+  }
+
+  static void UnregisterPlugin(
```

Comment:
Looks like this function is only used in the unit tests. I'm in favor of removing the function if possible. Less code is better.

Reply:
It was really helpful for making sure each test started in a clean state, but you're right, it's otherwise it's not necessary.
So, I derived from `AssetRegistry` in the unit tests and added the `UnregisterPlugin` method to that class instead. I think that's cleaner.